tutorials/lab/coral2.rst

```diff
@@ -9,75 +9,129 @@ El Capitan, and Frontier are similar in that they all use the
+have to specify ``-opmi=cray-pals`` may add the following lines near the top
+of the flux-shell's ``initrc.lua``, before any call to load plugins:
 
+.. code-block:: lua
 
-The lines should come before any call to load plugins.
+  if shell.options['pmi'] == nil then
+      shell.options['pmi'] = 'cray-pals,simple'
+  end
+
+This alters the system default of ``pmi=simple``, applies to all Flux
+instances, and has no effect if the user specifies ``-opmi=`` on the command
+line.  Note that ``simple`` is still the preferred way to bootstrap Flux
+itself, so it is advised for it to be retained in the default.
+
+--------------------------------
+Cray PMI Complications with Flux
+--------------------------------
+
+Cray PMI requires Flux to allocate two unique network port numbers to each
+multi-node job and communicate them via ``PMI_CONTROL_PORT`` in the job
+environment.  The Cray PMI library uses these ports to establish temporary
+network connections to exchange interconnect endpoints.  Two jobs sharing
+a node must not be allocated overlapping port numbers, or the jobs may fail.
+
+flux-coral2 supplies the ``cray_pals_port_distributor`` plugin to allocate
+a unique pair of ports per job.  However, each Flux instance has an
+independent allocator for the same port range, so a complication arises
+when multiple Flux instances are sharing a node's port space.  Therefore,
+when Cray PMI is in use, Flux instances *must not share nodes*.
+
+To minimize the possibility of batch jobs, which are fully independent Flux
+instances, handing out duplicate ports, it is recommended to configure
+node-exclusive scheduling for the top level resource manager on these
+systems.  This leaves no opportunity for conflicting port numbers to be
+assigned among the top-level batch jobs.  It doesn't protect against batch
+jobs scheduling Flux sub-instances that conflict, however.
+
+The port allocator defaults to using a pool of 1000 ports.  This places an
+upper limit of 500 on the number of concurrently executing multi-node jobs
+per Flux instance.  The system limit is much higher since each batch job
+is an independent Flux instance that can run many jobs.  Also, single node
+jobs do not consume a port pair and are not subject to this limit.
+
+--------------------------
+Troubleshooting Cray MPICH
+--------------------------
 
 If Flux jobs that use Cray MPICH end up as a collection of singletons,
-that is usually a sign that Cray MPICH is trying to use libPALS.
+or fail in ``MPI_Init()``, that is usually a sign that something is wrong
+in the PMI bootstrapping environment.  When this happens it may be useful to:
+
+- Add ``-o pmi=NAME[,NAME,...]`` to control which PMI implementations
+  are offered by the flux shell to jobs, e.g. ``simple``, ``cray-pals``,
+  ``pmix``).
+
+- Add ``-o verbose=2`` to request the shell to print tracing info
+  from the PMI implementations.
+
+- Launch ``flux-pmi`` as a parallel program to test PMI in isolation.
+
+.. code-block:: lua
```

Comment:
```suggestion
.. code-block:: console
```

Reply:
Does the lua tag cause a problem?  The lua syntax highlighting seems to be working in the preview...

Reply:
But isn’t it bash or console?

Reply:
I guess if you like it, there is no issue! Someone in the future might see it and change it though (if they think it’s not correct).

Reply:
Oops!  No no, you are correct, I was not zeroing in on the right code block.  Will fix!